webapp/src/cmds.ts

```diff
@@ -450,7 +464,128 @@ export function showDisconnectAsync(): Promise<void> {
 }
 
 export function disconnectAsync(): Promise<void> {
-    return pxt.packetio.disconnectAsync();
+    pxt.log("[CLIENT]: starting disconnect")
+    disconnectingPacketIO = true;
+    return pxt.packetio.disconnectAsync()
+        .then(() => {
+            pxt.log("[CLIENT]: sending confirmed disconnect " + lockRef)
+            hasLock = false;
+            sendServiceWorkerMessage({
+                type: "serviceworkerclient",
+                action: "release-packet-io-lock",
+                lock: lockRef
+            });
+            disconnectingPacketIO = false;
+        })
+}
+
+
+// Generate a unique id for communicating with the service worker
+const lockRef = pxtc.Util.guidGen();
+let pendingPacketIOLockResolver: () => void;
+let pendingPacketIOLockRejecter: () => void;
+let serviceWorkerSupportedResolver: () => void;
+let reconnectPromise: Promise<void>;
+let hasLock = false;
+let deployingPacketIO = false;
+let disconnectingPacketIO = false;
+let serviceWorkerSupported: boolean | undefined = undefined;
+
+async function requestPacketIOLockAsync() {
+    if (hasLock) return;
+    if (pendingPacketIOLockResolver) return Promise.reject("Already waiting for packet lock");
+    const supported = await checkIfServiceWorkerSupportedAsync();
+    if (!supported) return;
+
+    if (navigator?.serviceWorker?.controller) {
+        return new Promise<void>((resolve, reject) => {
```

Comment:
Is it good to mix async function and "raw" promises?

Reply:
I think you have to... promises have to come from somewhere, right?